syntax = "proto2";

package rpc;

message Request {
    optional uint64 id      = 1;
    optional string request = 2;
    optional string reply   = 3;
    optional bytes  body    = 4;
}

enum ErrorCode {
    NOERROR              = 0;  // success

    REQUEST_INVALID      = 1;  // Server: invalid request
    REQUEST_NOT_MATCHED  = 2;  // Server: request not matched
    REQUEST_PARSE_ERROR  = 3;  // Server: request parse failed
    REPLY_PACK_ERROR     = 10; // Server: reply pack failed

    REPLY_NOT_MATCHED    = 11; // Client: request not matched
    REPLY_PARSE_ERROR    = 12; // Client: reply parse failed
}

message Reply {
    optional uint64 id      = 1;
    optional string request = 2;
    optional string reply   = 3;
    optional bytes  body    = 4;
    optional ErrorCode errcode = 5;
}